File: lib/spread/valuation/pipeline.ts
// Orchestrates: cache -> eBay comps -> Claude appraisal -> auctioneer estimate fallback.
import type { Config } from "../config";
import type { Store } from "../store";
import type { Comp, Lot, Valuation } from "../types";
import { titleKey, usable, emptyValuation } from "./base";
import { ClaudeValuer } from "./claude";
import { cleanQuery, fetchActiveComps, fetchSoldComps, summarize } from "./ebay";
import { valueFromEstimate } from "./estimate";

const HOT_WORDS = /\b(dewalt|milwaukee|makita|bosch|snap-?on|festool|stihl|husqvarna|honda|yamaha|generac|kohler|john deere|kubota|caterpillar|toro|ego\b|ryobi|yeti|weber|traeger|big green egg|stanley|bailey|griswold|wagner|rolex|omega|seiko|tag heuer|breitling|cartier|tiffany|bulova|accutron|elgin|waltham|hamilton|pocket watch|gold|silver|sterling|platinum|diamond|karat|\d+k\b|gold filled|bullion|coin|morgan|eagle|krugerrand|cameo|bakelite|trifari|weiss|coro|monet|napier|costume jewelry|louis vuitton|gucci|coach|prada|chanel|hermes|dooney|lego|nintendo|playstation|ps5|xbox|switch|pokemon|magic the gathering|funko|matchbox|hot wheels|lionel|marx|tonka|buddy l|barbie|g\.?i\.? joe|star wars|pez|topps|bowman|fleer|upper deck|panini|donruss|psa|bgs|rookie|autograph|signed|baseball card|football card|sports card|basketball card|hockey card|trading card|wax pack|apple|iphone|ipad|macbook|imac|samsung|sony|canon|nikon|leica|kodak|polaroid|brownie|bose|sonos|dyson|kitchenaid|vitamix|gibson|fender|martin|taylor|roland|marshall|zenith|philco|typewriter|rotary phone|herman miller|aeron|steelcase|eames|knoll|mid.?century|art deco|art nouveau|victorian|primitive|folk art|occupied japan|noritake|kewpie|bisque|chalkware|hummel|lladro|roseville|mccoy|hull|fenton|fiesta|depression glass|carnival glass|milk glass|wedgwood|jasperware|lenox|royal doulton|limoges|haviland|stoneware|crock|redware|ironstone|pyrex|corning|majolica|cloisonne|satsuma|imari|nippon|wall pocket|salt and pepper|shakers|figurine|advertising|tin sign|porcelain sign|coca.?cola|pepsi|still bank|mechanical bank|calendar bank|letter opener|inkwell|fountain pen|parker|sheaffer|waterman|zippo|lighter|pocket knife|case xx|buck knife|humidor|tobacco|oil lamp|aladdin|kerosene|lantern|railroad|insulator|marbles|license plate|milk bottle|whiskey bottle|decanter|jim beam|ezra brooks|singer|featherweight|seth thomas|ansonia|clock|first edition|1st edition|antique book|primer|mcguffey|yearbook|program|pennant|comic|boy scouts|bsa\b|girl scouts|chip hilton|hardy boys|nancy drew|big little book|lp\b|vinyl|record|45 rpm|78 rpm|vintage|antique|rare|native|indian|wall hanging|trek|specialized|cannondale|giant|peloton|schwinn|garmin|dji|gopro|oculus|quest)\b/gi;

let VISION = true; // set from config by ValuationPipeline

/** Cheap heuristic: which lots are worth spending a valuation call on, highest first. */
export function triageScore(lot: Lot): number {
  const title = lot.title || "";
  if (title.length < 6) return -1;
  let s = Math.min(3, (title.match(HOT_WORDS) ?? []).length);
  if (/\b[A-Z]{1,4}\d{2,5}[A-Z]?\b/.test(title)) s += 1.5;
  if ((lot.bid_count || 0) === 0) s += 1;
  if (lot.estimate) s += 0.5;
  if (lot.time_left_seconds !== null && lot.time_left_seconds < 6 * 3600) s += 1;
  if (/\b(box of|misc|assorted|miscellaneous|contents of|shelf lot|knic ?knac|knick ?knack|bric.a.brac|smalls)\b/i.test(title))
    s += (lot.picture_count ?? 0) >= 1 && VISION ? 0.5 : -1.5; // with photos, vague titles are where unnoticed value hides
  if ((lot.min_bid || lot.high_bid || 0) <= 5) s += 0.5; // penny lots are the bread and butter
  return s;
}

export class ValuationPipeline {
  private claude: ClaudeValuer | null | undefined;
  constructor(private c: Config, private store: Store, claude?: ClaudeValuer | null, private pictureFetcher?: (lot: Lot) => Promise<string[]>) {
    this.claude = claude;
    VISION = c.vision;
  }

  private getClaude(): ClaudeValuer | null {
    if (this.claude !== undefined) return this.claude;
    this.claude = this.c.claudeEnabled ? new ClaudeValuer(this.c.model, this.c.webSearch, 3, this.c.vision, this.c.maxImages) : null;
    return this.claude;
  }

  async valueLot(lot: Lot, force = false): Promise<Valuation> {
    const key = titleKey(lot.title, lot.quantity);
    if (!force) {
      const cached = await this.store.cachedValuation(key, this.c.valuationTtlDays * 86400);
      if (cached) {
        const v = { ...cached, lot_id: lot.id, cache_hit: true };
        await this.store.saveValuation(lot.id, v);
        return v;
      }
    }
    let comps: Comp[] = [];
    if (this.c.valuer !== "none" && this.c.valuer !== "estimate") {
      const q = cleanQuery(lot.title);
      if (this.c.ebaySold) comps = await fetchSoldComps(q);
      if (this.c.ebayClientId && this.c.ebayClientSecret) comps = comps.concat(await fetchActiveComps(q, this.c.ebayClientId, this.c.ebayClientSecret));
    }
    let val: Valuation | null = null;
    const claude = this.c.valuer === "auto" || this.c.valuer === "claude" ? this.getClaude() : null;
    if (claude) {
      if (this.c.vision && !lot.pictures?.length && this.pictureFetcher) {
        try {
          const pics = await this.pictureFetcher(lot);
          if (pics.length) lot = { ...lot, pictures: pics };
        } catch (e) {
          console.info("picture fetch failed for", lot.id, e);
        }
      }
      val = await claude.value(lot, comps);
      if (!usable(val)) val = null;
    }
    if (!val && (this.c.valuer === "auto" || this.c.valuer === "ebay")) val = this.fromComps(lot, comps, key);
    if (!val) {
      val = valueFromEstimate(lot);
      if (comps.length) val.comps = comps.slice(0, 10);
    }
    val.created_at = Date.now() / 1000;
    await this.store.saveValuation(lot.id, val);
    return val;
  }

  private fromComps(lot: Lot, comps: Comp[], key: string): Valuation | null {
    const sold = comps.filter((c) => c.source === "ebay_sold");
    const pool = sold.length >= 3 ? sold : comps;
    const s = summarize(pool);
    if (s.n < 3) return null;
    const v = emptyValuation(lot);
    v.title_key = key;
    v.identified_item = lot.title;
    v.low = s.p25!; v.mid = s.median!; v.high = s.p75!;
    const active = pool === comps && sold.length < 3;
    v.confidence = Math.min(active ? 0.5 : 0.7, 0.3 + 0.04 * s.n) * (active ? 0.8 : 1);
    v.method = active ? "ebay_active" : "ebay_sold";
    v.confidence_reason = `${active ? "Median asking price of" : "Median of"} ${s.n} eBay ${active ? "active" : "sold"} listings matching the title (unverified item match).`;
    v.comps = pool.slice(0, 15);
    v.rationale = `eBay ${active ? "active" : "sold"} comps for '${cleanQuery(lot.title)}': median $${Math.round(s.median!)}, IQR $${Math.round(s.p25!)}–$${Math.round(s.p75!)} across ${s.n} listings.`;
    v.risks = ["Comps matched on title keywords only; confirm the exact model/condition before bidding."];
    if (active) v.risks.push("Asking prices, not sold prices: expect realized values 10–30% lower.");
    v.search_query = cleanQuery(lot.title);
    return v;
  }

  /** Value the most promising lots, `maxLots` at most, stopping when `deadline` (ms epoch) passes. */
  async valueMany(lots: Lot[], opts: { maxLots?: number; deadline?: number; onDone?: (n: number, total: number, lot: Lot) => void } = {}): Promise<number> {
    let todo = lots.filter((l) => !l.is_closed && triageScore(l) >= 0).sort((a, b) => triageScore(b) - triageScore(a));
    if (opts.maxLots !== undefined) todo = todo.slice(0, Math.max(0, opts.maxLots));
    let idx = 0, done = 0;
    const worker = async () => {
      while (idx < todo.length) {
        if (opts.deadline && Date.now() > opts.deadline) return;
        const lot = todo[idx++];
        try {
          await this.valueLot(lot);
        } catch (e) {
          console.warn("valuation crashed for lot", lot.id, e);
        }
        done++;
        opts.onDone?.(done, todo.length, lot);
      }
    };
    await Promise.all(Array.from({ length: Math.min(this.c.valuationWorkers, todo.length) }, worker));
    return done;
  }
}

File: tools/auction-arbitrage/arb/valuation/pipeline.py
"""Orchestrates: cache -> eBay sold comps -> Claude appraisal -> auctioneer estimate fallback."""
from __future__ import annotations

import logging
import re
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import Any, Callable, Iterable

from ..config import Settings
from ..db import Store
from .base import Comp, Valuation, title_key
from .ebay import clean_query, fetch_sold_comps
from .estimate import value_from_estimate

log = logging.getLogger(__name__)

# Words that hint at resale-grade items; used only to prioritise which lots get valued first.
_HOT_WORDS = re.compile(
    r"\b(dewalt|milwaukee|makita|bosch|snap-?on|festool|stihl|husqvarna|honda|yamaha|generac|kohler|"
    r"rolex|omega|seiko|tag heuer|breitling|cartier|tiffany|louis vuitton|gucci|coach|prada|chanel|hermes|"
    r"gold|silver|sterling|platinum|diamond|karat|\d+k\b|bullion|coin|morgan|eagle|krugerrand|"
    r"lego|nintendo|playstation|ps5|xbox|switch|pokemon|magic the gathering|funko|"
    r"apple|iphone|ipad|macbook|imac|samsung|sony|canon|nikon|leica|bose|sonos|dyson|kitchenaid|vitamix|"
    r"herman miller|aeron|steelcase|eames|knoll|mid.?century|"
    r"yeti|weber|traeger|big green egg|"
    r"gibson|fender|martin|taylor|yamaha|roland|marshall|"
    r"trek|specialized|cannondale|giant|peloton|"
    r"john deere|kubota|cat\b|caterpillar|toro|ego\b|ryobi|"
    r"garmin|dji|gopro|oculus|quest|"
    r"vintage|antique|signed|first edition|rare)\b", re.I)


def triage_score(lot: dict[str, Any]) -> float:
    """Cheap heuristic: which lots are worth spending a valuation call on, highest first."""
    title = lot.get("title") or ""
    s = 0.0
    if len(title) < 6:
        return -1.0
    s += min(3, len(_HOT_WORDS.findall(title)))
    if re.search(r"\b[A-Z]{1,4}\d{2,5}[A-Z]?\b", title):  # looks like a model number
        s += 1.5
    if (lot.get("bid_count") or 0) == 0:
        s += 1.0  # nobody has noticed it yet
    if lot.get("estimate"):
        s += 0.5
    tl = lot.get("time_left_seconds")
    if tl is not None and tl < 6 * 3600:
        s += 1.0
    if re.search(r"\b(box of|misc|assorted|miscellaneous|contents of|shelf lot)\b", title, re.I):
        s -= 1.5
    return s


class ValuationPipeline:
    def __init__(self, settings: Settings, store: Store, *, claude_valuer: Any | None = None):
        self.settings = settings
        self.store = store
        self._claude = claude_valuer
        self._claude_tried = claude_valuer is not None

    # ---------------------------------------------------------- provider setup
    def _get_claude(self):
        if self._claude_tried:
            return self._claude
        self._claude_tried = True
        if self.settings.valuer in ("none", "ebay", "estimate"):
            return None
        if not self.settings.anthropic_available:
            log.warning("ANTHROPIC_API_KEY not set: falling back to eBay comps / auctioneer estimates only")
            return None
        try:
            from .claude import ClaudeValuer
            self._claude = ClaudeValuer(self.settings.model, web_search=self.settings.web_search)
        except Exception as e:  # SDK missing, bad key format, etc.
            log.warning("Claude valuer unavailable: %s", e)
            self._claude = None
        return self._claude

    # -------------------------------------------------------------- one lot
    def value_lot(self, lot: dict[str, Any], *, force: bool = False) -> Valuation:
        key = title_key(lot.get("title", ""), lot.get("quantity"))
        if not force:
            cached = self.store.cached_valuation(key, self.settings.valuation_ttl_days * 86400)
            if cached:
                cached = dict(cached, lot_id=lot["id"], cache_hit=True)
                self.store.save_valuation(lot["id"], cached)
                return Valuation(**{k: v for k, v in cached.items() if k in Valuation.__dataclass_fields__})

        comps: list[Comp] = []
        if self.settings.ebay_sold and self.settings.valuer in ("auto", "ebay", "claude"):
            comps = fetch_sold_comps(clean_query(lot.get("title", "")))

        val: Valuation | None = None
        claude = self._get_claude() if self.settings.valuer in ("auto", "claude") else None
        if claude is not None:
            val = claude.value(lot, comps)
            if not val.usable:
                log.info("claude gave no usable value for %s (%s)", lot["id"], val.error or val.rationale[:80])
                val = None

        if val is None and comps and self.settings.valuer in ("auto", "ebay"):
            val = self._from_comps(lot, comps, key)

        if val is None:
            val = value_from_estimate(lot)
            if comps:
                val.comps = [c.__dict__ for c in comps[:10]]

        val.created_at = time.time()
        d = val.to_dict()
        self.store.save_valuation(lot["id"], d)
        return val

    @staticmethod
    def _from_comps(lot: dict[str, Any], comps: list[Comp], key: str) -> Valuation | None:
        from .ebay import summarize
        s = summarize(comps)
        if s.get("n", 0) < 3:
            return None
        v = Valuation(lot_id=lot["id"], title_key=key, identified_item=lot.get("title", ""))
        v.low, v.mid, v.high = s["p25"], s["median"], s["p75"]
        v.confidence = min(0.7, 0.3 + 0.04 * s["n"])
        v.confidence_reason = f"Median of {s['n']} eBay sold listings matching the title (unverified item match)."
        v.method = "ebay_sold"
        v.comps = [c.__dict__ for c in comps[:15]]
        v.rationale = (f"eBay sold comps for '{clean_query(lot.get('title', ''))}': median ${s['median']:,.0f}, "
                       f"IQR ${s['p25']:,.0f}–${s['p75']:,.0f} across {s['n']} sales.")
        v.risks = ["Comps matched on title keywords only; confirm the exact model/condition before bidding."]
        v.search_query = clean_query(lot.get("title", ""))
        return v

    # ------------------------------------------------------------ many lots
    def value_many(self, lots: Iterable[dict[str, Any]], *, max_lots: int | None = None,
                   progress: Callable[[int, int, dict[str, Any]], None] | None = None) -> int:
        todo = [l for l in lots if not l.get("is_closed")]
        todo.sort(key=triage_score, reverse=True)
        todo = [l for l in todo if triage_score(l) >= 0]
        if max_lots:
            todo = todo[:max_lots]
        done = 0
        workers = max(1, self.settings.valuation_workers)
        with ThreadPoolExecutor(max_workers=workers) as ex:
            futs = {ex.submit(self.value_lot, lot): lot for lot in todo}
            for fut in as_completed(futs):
                lot = futs[fut]
                try:
                    fut.result()
                except Exception as e:  # never let one lot kill the scan
                    log.exception("valuation crashed for lot %s: %s", lot.get("id"), e)
                done += 1
                if progress:
                    progress(done, len(todo), lot)
        return done
